### PR TITLE
CSHARP-4298: SDAM should prioritize electionId over setVersion only on >=6.0 servers

### PR DESCRIPTION
**Pull request by BorisDog**

### CODE REVIEW

File: src/MongoDB.Driver.Core/Core/Clusters/MultiServerCluster.cs [hunk 1]
```diff
@@ -676,51 +645,88 @@ private static class State
...
                 }
-                if (_setVersion > setVersion)
+                else
                 {
-                    return false;
+                    if (setVersion != null && electionId != null)
+                    {
+                        if (_setVersion != null &&
+                            _electionId != null &&
+                            _setVersion > setVersion || _setVersion == setVersion && Compare(electionId) > 0)
```
Comment (DmitryLukyanov) [inline src/MongoDB.Driver.Core/Core/Clusters/MultiServerCluster.cs:689]: let's add brackets in the same way as it's mentioned in the pseudocode: https://github.com/mongodb/specifications/commit/37c32cf5200cd8da6652c75ac65279c43628fe7a#diff-52a48d3b84e9e90e30fb87b3799c6302750f1ed2817b2abbead44e7a7bb6bb47R1136

File: src/MongoDB.Driver.Core/Core/Clusters/MultiServerCluster.cs [hunk 2]
```diff
@@ -676,51 +645,88 @@ private static class State
...
+                        {
+                            isNewPrimaryStale = true;
```
Comment (DmitryLukyanov) [inline src/MongoDB.Driver.Core/Core/Clusters/MultiServerCluster.cs:691]: should it be return here (based on): https://github.com/mongodb/specifications/commit/37c32cf5200cd8da6652c75ac65279c43628fe7a#diff-52a48d3b84e9e90e30fb87b3799c6302750f1ed2817b2abbead44e7a7bb6bb47R1145 ?

File: src/MongoDB.Driver.Core/Core/Clusters/MultiServerCluster.cs [hunk 3]
```diff
@@ -676,51 +645,88 @@ private static class State
...
-                /* above is equivalent to:
-                 * return
-                 *   _setVersion < setVersion
-                 *   || _setVersion == setVersion && (_electionId == null || _electionId.CompareTo(electionId) < 0); */
+                return _electionId.CompareTo(electionId);
             }
+
+            public override int GetHashCode() => base.GetHashCode();
+
+            private bool IsGreater(int? left, int? right) =>
```
Comment (DmitryLukyanov) [inline src/MongoDB.Driver.Core/Core/Clusters/MultiServerCluster.cs:725]: I would suggest adding a comment why we consider `left:X` bigger than `null`

File: src/MongoDB.Driver.Core/Core/Clusters/MultiServerCluster.cs [hunk 4]
```diff
@@ -344,105 +344,74 @@ private ClusterDescription ProcessReplicaSetChange(ClusterDescription clusterDes
...
+                                _maxElectionInfo.SetVersion,
+                                _maxElectionInfo.ElectionId));
+
+                            return clusterDescription.WithServerDescription(
+                                new ServerDescription(server.ServerId, server.EndPoint, "ReportedPrimaryIsStale"));
                         }
                     }
 
-                    if (isCurrentPrimaryStale)
+                    if (!_maxElectionInfo.Equals(newMaxElectionInfo))
```
[the remark this answers is not in the document]
Comment (BorisDog) [reply src/MongoDB.Driver.Core/Core/Clusters/MultiServerCluster.cs:380]: It's equivalent:
Before: `isCurrentPrimaryStale == true` IFF _maxElectionInfo was not set or primary is stale
Now: `!_maxElectionInfo.Equals(newMaxElectionInfo)` IFF _maxElectionInfo was not set (default) or primary is stale

File: src/MongoDB.Driver.Core/Core/Clusters/MultiServerCluster.cs (hunk 1 above)
Comment (BorisDog) [reply src/MongoDB.Driver.Core/Core/Clusters/MultiServerCluster.cs]: Done.

File: src/MongoDB.Driver.Core/Core/Clusters/MultiServerCluster.cs (hunk 2 above)
Comment (BorisDog) [reply src/MongoDB.Driver.Core/Core/Clusters/MultiServerCluster.cs]: Yes, isNewPrimaryStale is set, and return happens in level above.
But refactored this for better clarity.

File: src/MongoDB.Driver.Core/Core/Clusters/MultiServerCluster.cs (hunk 3 above)
Comment (BorisDog) [reply src/MongoDB.Driver.Core/Core/Clusters/MultiServerCluster.cs:726]: Done.

File: src/MongoDB.Driver.Core/Core/Servers/DefaultServer.cs [hunk 5]
```diff
@@ -173,9 +173,12 @@ protected override void Invalidate(string reasonInvalidated, bool clearConnectio
                     lastUpdateTimestamp: DateTime.UtcNow,
                     topologyVersion: topologyVersion);
 
+            var (host, port) = ServerId.EndPoint.GetHostAndPort();
```
[the remark this answers is not in the document]
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Servers/DefaultServer.cs:176]: I think it was done incorrectly, should we change it now (all this internal)? Or we can create a ticket for this and take later with some other logging work

File: src/MongoDB.Driver.Core/Core/Clusters/MultiServerCluster.cs (hunk 2 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Clusters/MultiServerCluster.cs]: I would say better clarity is to be able to map code in the spec and on our side, at least I wasn't able to do it initially. But I will look again

File: src/MongoDB.Driver.Core/Core/Servers/DefaultServer.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (BorisDog) [reply src/MongoDB.Driver.Core/Core/Servers/DefaultServer.cs:176]: There is not assumption that the endpoint is DnsEndpoint, host and port can be null.
It's debatable where this part should be:
1. We don't have any convince logs methods for logging endpoints yet, probably we'll consider that when 
2. Default server decides what it wants to log here, it logs DnsEndpoint if it is a DnsEndpoint.

I think we should revisit this if/when adding more non-spec logs using endpoints.

[the remark this answers is not in the document]
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Servers/DefaultServer.cs:176]: I think better to stick to the logic we have in other places like here: https://github.com/mongodb/mongo-csharp-driver/blob/master/src/MongoDB.Driver.Core/Core/Authentication/GssapiAuthenticator.cs#L238

[the remark this answers is not in the document]
Comment (BorisDog) [reply src/MongoDB.Driver.Core/Core/Servers/DefaultServer.cs:176]: Good point re. `IPEndpoint`, missed that previously. Done.

File: src/MongoDB.Driver.Core/Core/Clusters/MultiServerCluster.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Clusters/MultiServerCluster.cs:380]: This wording makes hard to follow what happens here. `!Equals` doesn't mean stale, it simply means different. I think we need a better naming here that will allow better understand what happens without deep dive into impl details. For example similar as it was before: `IsStale` and so on. See what you think about it

[the remark this answers is not in the document]
Comment (BorisDog) [reply src/MongoDB.Driver.Core/Core/Clusters/MultiServerCluster.cs:380]: The stale check happens in `ElectionInfo.Compare`. So `Equals` is a simple comparison, and update happens only if `newMaxElectionInfo` is different than current one.

File: src/MongoDB.Driver.Core/Core/Misc/EndPointHelper.cs [hunk 6]
```diff
@@ -158,7 +158,12 @@ public static EndPoint Parse(string value)
         }
 
         internal static (string Host, int Port) GetHostAndPort(this EndPoint endPoint) =>
-            (endPoint is DnsEndPoint dnsEndPoint) ? (dnsEndPoint.Host, dnsEndPoint.Port) : default;
+            endPoint switch
+            {
+                DnsEndPoint dnsEndPoint => (dnsEndPoint.Host, dnsEndPoint.Port),
```
[the remark this answers is not in the document]
Comment (BorisDog) [reply src/MongoDB.Driver.Core/Core/Misc/EndPointHelper.cs:163]: Yeah we should introduce more helpers on logger level for more convenient logs handling, that are not part of spec.
It's just a bit more complicated so I'd prefer to postpone this until we have a bit more of those.

**Review by DmitryLukyanov — APPROVED** [review]
LGTM (there is a question we can discuss if you want, but I'm ok with it as is)

File: src/MongoDB.Driver.Core/Core/Clusters/MultiServerCluster.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (BorisDog) [reply src/MongoDB.Driver.Core/Core/Clusters/MultiServerCluster.cs:380]: Yeah this is exactly this logic: maxElectionInfo update is equivalent/definition of current primary staleness (otherwise it would not be updated).

**Comment by JamesKovacs** [comment]
We have 5 failing variants in this patch build. Is there any commonality between the failures? Have these changes destabilized certain tests?

**Comment by DmitryLukyanov** [comment]
Please let me know what you will find in EG results
